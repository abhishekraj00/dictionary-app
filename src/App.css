* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

.themeIconContainer {
  margin-bottom: 10%;
}
