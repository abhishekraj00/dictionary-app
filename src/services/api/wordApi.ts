export const randomWordApi = "https://random-word-api.herokuapp.com/word";
export const dictionaryWordApi =
  "https://api.dictionaryapi.dev/api/v2/entries/en/";
