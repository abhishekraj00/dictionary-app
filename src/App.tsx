import "./App.css";
import styled, { ThemeProvider } from "styled-components";
import { ReactComponent as DarkIcon } from "../src/utils/icons/dark.svg";
import { ReactComponent as LightIcon } from "../src/utils/icons/light.svg";
import { darkTheme, lightTheme } from "./utils/theme";
import { useEffect, useState } from "react";
import RandomWord from "./components/RandomWord";
import Search from "./components/Search/Search";
import { dictionaryWordApi } from "./services/api/wordApi";
import { getRandomWord } from "./utils/helper/helpers";
import { words } from "./utils/randomWord";
import { Modal } from "./components/Modal";

const Body = styled.div`
  background: ${(props) => props.theme.background};
  color: ${(props) => props.theme.color};
  margin-top: 5%;
  display: flex;
  justify-content: center;
  align-items: center;
`;
const Container = styled.div`
  background: ${(props) => props.theme.background};
  color: ${(props) => props.theme.color};
  padding: 1rem;
  width: 600px;
  border: 1px solid black;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  border-radius: 1rem;
`;

const ModalButton = styled.button`
  border-radius: 1rem;
  background: ${(props) => props.theme.color};
  color: ${(props) => props.theme.background};
`;

// Interface of definitions
interface definitions {
  definition: string;
  example: string;
}
//Interface of meaning data
interface meaningDataApi {
  definitions: definitions[];
}

//Interface of word data
export interface wordApiData {
  word: string;
  meanings: meaningDataApi[];
}

const App = () => {
  const [mode, setMode] = useState<boolean>(false);
  const [searchWord, setSearchWord] = useState<string>("");
  const [randomWord, setRandomWord] = useState<wordApiData | null>(null);
  const [showModal, setShowModal] = useState<boolean>(false);
  const [errorMessage, setErrorMessage] = useState("");

  //random word only generate while in mounting face
  useEffect(() => {
    let finalWord = searchWord ? searchWord : getRandomWord(words);
    fetch(dictionaryWordApi + finalWord)
      .then((response) => {
        if (response.ok) {
          return response.json();
        } else {
          throw new Error("oops! Word Not Found ðŸ˜¥");
        }
      })
      .then((data) => {
        setRandomWord(data[0]);
      })
      .catch((error) => {
        setErrorMessage(error.message);
        setRandomWord(null);
      });
  }, [searchWord, setRandomWord]);

  const handleOpenModal = () => {
    setShowModal(true);
  };

  return (
    <>
      <Body>
        <ThemeProvider theme={mode ? darkTheme : lightTheme}>
          <Container data-testid="container">
            <div className="themeIconContainer">
              {mode ? (
                <LightIcon
                  width={25}
                  height={25}
                  onClick={() => setMode(!mode)}
                  data-testid="light-icon"
                />
              ) : (
                <DarkIcon
                  width={25}
                  height={25}
                  onClick={() => setMode(!mode)}
                  data-testid="dark-icon"
                />
              )}
            </div>
            <Search setSearchWord={setSearchWord} />
            <RandomWord randomWord={randomWord} errorMessage={errorMessage} />
            <ModalButton className="btn btn-dark" onClick={handleOpenModal}>
              Show More
            </ModalButton>
            <Modal
              randomWord={randomWord}
              showModal={showModal}
              setShowModal={setShowModal}
            />
          </Container>
        </ThemeProvider>
      </Body>
    </>
  );
};

export default App;
