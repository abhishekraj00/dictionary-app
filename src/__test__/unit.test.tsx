import { render, screen, fireEvent } from "@testing-library/react";
import { dictionaryWordApi } from "../services/api/wordApi";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import App from "../App";
import "@testing-library/jest-dom/extend-expect";

describe("App component sanpshot test", () => {
  test("should render correctly", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("App component featcher test", () => {
  test("renders light theme by default", () => {
    render(<App />);
    const container = screen.getByTestId("container");
    expect(container).toHaveStyle("background: #fff");
    expect(container).toHaveStyle("color: #333");
  });

  test("switches to dark theme when dark icon is clicked", () => {
    render(<App />);
    const darkIcon = screen.getByTestId("dark-icon");
    fireEvent.click(darkIcon);
    const container = screen.getByTestId("container");
    expect(container).toHaveStyle("background: #333");
    expect(container).toHaveStyle("color: #fff");
  });

  test("should return word data when a valid word is provided", async () => {
    const response = await fetch(`${dictionaryWordApi}hello`);
    const data = await response.json();

    expect(response.ok).toBe(true);
    expect(Array.isArray(data)).toBe(true);
    expect(data.length).toBeGreaterThan(0);
    expect(data[0]).toHaveProperty("word");
    expect(data[0]).toHaveProperty("meanings");
  });
  test("renders error message when API response fails", async () => {
    const response = await fetch(`${dictionaryWordApi}asdas`);
    await response.json();

    expect(response.ok).toBe(false);
  });

  test("opens and closes the modal when clicking the Show More button", () => {
    render(<App />);
    const showModalButton = screen.getByText("Show More");

    fireEvent.click(showModalButton);
    expect(screen.getByTestId("modal")).toBeInTheDocument();

    fireEvent.click(screen.getByTestId("modal-close"));
    expect(screen.queryByTestId("modal")).not.toBeInTheDocument();
  });

  test("displays an error message when the word is not found", async () => {
    global.fetch = jest.fn(() =>
      Promise.reject(new Error("oops! Word Not Found ðŸ˜¥"))
    );

    render(<App />);
    const searchInput = screen.getByPlaceholderText("Search a word...");
    fireEvent.change(searchInput, { target: { value: "nonexistent" } });

    const errorMessage = await screen.findByTestId("error-message");
    expect(errorMessage).toHaveTextContent("oops! Word Not Found ðŸ˜¥");
  });
});
