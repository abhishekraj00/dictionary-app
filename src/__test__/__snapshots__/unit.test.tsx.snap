// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component sanpshot test should render correctly 1`] = `
<div
  className="sc-gLLvby gmcYa"
>
  <div
    className="sc-iBdmCd jqISTi"
    data-testid="container"
  >
    <div
      className="themeIconContainer"
    >
      <svg
        data-testid="dark-icon"
        height={25}
        onClick={[Function]}
        width={25}
      >
        dark.svg
      </svg>
    </div>
    <div
      className="sc-hLQRIN eIjcEg"
    >
      <input
        className="sc-eDLJxc bbZwxl"
        onChange={[Function]}
        onKeyPress={[Function]}
        placeholder="Search a word..."
        type="text"
        value=""
      />
      <svg
        height={20}
        width={20}
      >
        search.svg
      </svg>
    </div>
    <div
      className="sc-beySbM dYDsNu"
    >
      <h2
        data-testid="definition"
      >
        Word: 
        <h4
          className="sc-guDLRT jlXoGf"
        >
          <h4
            className="sc-dmyDGi ksKYvN"
          >
            Loading...
          </h4>
        </h4>
      </h2>
    </div>
    <button
      className="sc-fsYeqs LApSS btn btn-dark"
      onClick={[Function]}
    >
      Show More
    </button>
  </div>
</div>
`;
